File: backend/src/main/java/com/zihler/wiki/domain/values/ReferencedWikiPages.java
package com.zihler.wiki.domain.values;

import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

public class ReferencedWikiPages {
    private Set<ReferenceTag> referencedWikiPages;

    private ReferencedWikiPages() {
        referencedWikiPages = new HashSet<>();
    }

    private ReferencedWikiPages(Set<ReferenceTag> referencedWikiPages) {
        this.referencedWikiPages = referencedWikiPages;
    }

    public static ReferencedWikiPages from(Set<ReferenceTag> referencedWikiPages) {
        return new ReferencedWikiPages(referencedWikiPages);
    }

    public static ReferencedWikiPages empty() {
        return new ReferencedWikiPages();
    }

    public void add(ReferenceTag referencedWikiPage) {
        referencedWikiPages.add(referencedWikiPage);
    }

    public Set<ReferenceTag> getReferencedWikiPages() {
        return referencedWikiPages;
    }

    @Override
    public String toString() {
        return "\"referencedWikiPages\":" + referencedWikiPages;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        ReferencedWikiPages that = (ReferencedWikiPages) o;
        return Objects.equals(referencedWikiPages, that.referencedWikiPages);
    }

    @Override
    public int hashCode() {
        return Objects.hash(referencedWikiPages);
    }
}

File: frontend/src/app/wiki/domain/referenced-wiki-pages.ts
import {ReferenceTag} from './reference-tag';

export class ReferencedWikiPages {
  constructor(private _referencedWikiPages: ReferenceTag[] = []) {
  }

  get referencedWikiPages(): ReferenceTag[] {
    return this._referencedWikiPages;
  }

  set referencedWikiPages(value: ReferenceTag[]) {
    this._referencedWikiPages = value;
  }

  static from(referencedWikiPages: string[]) {
    return new ReferencedWikiPages(referencedWikiPages.map(referenceTagString => ReferenceTag.from(referenceTagString)));
  }

  static empty() {
    return new ReferencedWikiPages();
  }
}
